docs/SIL.rst

```diff
@@ -1210,17 +1210,36 @@ Global Variables
 ::
 
   decl ::= sil-global-variable
+  static-initializer ::= '=' '{' sil-instruction-def* '}'
   sil-global-variable ::= 'sil_global' sil-linkage identifier ':' sil-type
+                             (static-initializer)?
 
 SIL representation of a global variable.
 
-Global variable access is performed by the ``alloc_global`` and ``global_addr``
-SIL instructions. Prior to performing any access on the global, the
-``alloc_global`` instruction must be performed to initialize the storage.
+Global variable access is performed by the ``alloc_global``, ``global_addr``
+and ``global_value`` instructions.
 
+A global can have a static initializer if it's initial value can be
+composed of literals. The static initializer is represented as a list of
+literal and aggregate instructions where the last instruction is the top-level
+value of the static initializer::
+
+  sil_global hidden @_T04test3varSiv : $Int {
+    %0 = integer_literal $Builtin.Int64, 27
+    %initval = struct $Int (%0 : $Builtin.Int64)
+  }
+
+If a global does not have a static initializer, the ``alloc_global``
+instruction must be performed prior an access to initialize the storage.
 Once a global's storage has been initialized, ``global_addr`` is used to
 project the value.
 
+If the last instruction in the static initializer is an ``object`` instruction
+the global variable is a statically initialized object. In this case the
+variable cannot be used as l-value, i.e. the reference to the object cannot be
+modified. As a consequence the variable cannot be accessed with ``global_addr``
+but only with ``global_value``.
```

Comment:
Fair enough. Another thing that crossed my mind—is the constraint on `object` strictly necessary? It seems like something we could support generally by lowering to alloc_ref + initialization of the class fields during IRGen, and it might make some analyses easier to be able to represent object instantiations that way.